### ISSUE TITLE
Node w/ failed instance health checks slow replacement

### ISSUE THREAD

**Issue by ccassidy-maxar** [aws/aws-parallelcluster#2340]
**Required Info:**
 - AWS ParallelCluster version 2.10.0
 - Full cluster configuration without any credentials or personal data
```
[cluster AZ-a]
key_name = [redacted]
base_os = alinux2
master_instance_type = r5.24xlarge
vpc_settings = AZ-a
disable_cluster_dns = false
master_root_volume_size = 3000
compute_root_volume_size = 30
fsx_settings = fs
queue_settings = [redacted]
ebs_settings = gp2
cw_log_settings = default
dashboard_settings = default
ec2_iam_role = [redacted]
scheduler = slurm
post_install = [redacted]
s3_read_resource = [redacted]

[vpc AZ-a]
vpc_id = [redacted]
master_subnet_id = [redacted]
use_public_ips = false
vpc_security_group_id = [redacted]

[fsx fs]
shared_dir = [redacted]
storage_capacity = 14400
imported_file_chunk_size = 1024
import_path = [redacted]
deployment_type = SCRATCH_2

[ebs gp2]
shared_dir = /shared
volume_type = gp2
volume_size = 3000

[cw_log ParallelCluster]
enable = false
retention_days = 14

[aliases]
ssh = ssh {CFN_USER}@{MASTER_IP} {ARGS}
```

**Bug description and how to reproduce:**
We spun up a cluster with 2.10.0 today and one of the compute nodes failed to initialize and started failing the instance reachability checks a few minutes after launch,  The clustermgtd logs indicate that it saw the node failing health checks and set it to 'DRAIN'.  It kept indicating that it was in the same DRAIN state for about 50 minutes before it set to 'DRAINED' and terminated/replaced the instance.  With prior versions on ASGs when we saw a node fail health checks it was typically replaced within minutes.  Is this longer delay expected with the new version and if so are there any settings we can tweak to make it more sensitive to instance health checks?

** Logs **
WARNING events from clustermgtd:
```
2020-12-28 15:20:00,893 - [slurm_plugin.clustermgtd:_handle_health_check] - WARNING - Node st-c518xlarge-148(10.230.131.146) is associated with instance i-0c52326aac5dd766b that is failing ec2_health_check. EC2 health state: EC2InstanceHealthState(id='i-0c52326aac5dd766b', state='running', instance_status={'Details': [{'ImpairedSince': datetime.datetime(2020, 12, 28, 15, 12, tzinfo=tzlocal()), 'Name': 'reachability', 'Status': 'failed'}], 'Status': 'impaired'}, system_status={'Details': [{'Name': 'reachability', 'Status': 'passed'}], 'Status': 'ok'}, scheduled_events=None)
2020-12-28 15:20:00,893 - [slurm_plugin.clustermgtd:_handle_health_check] - WARNING - Setting nodes failing health check type ec2_health_check to DRAIN: ['st-c518xlarge-148']
2020-12-28 15:21:01,369 - [slurm_plugin.clustermgtd:_handle_health_check] - WARNING - Node st-c518xlarge-148(10.230.131.146) is associated with instance i-0c52326aac5dd766b that is failing ec2_health_check. EC2 health state: EC2InstanceHealthState(id='i-0c52326aac5dd766b', state='running', instance_status={'Details': [{'ImpairedSince': datetime.datetime(2020, 12, 28, 15, 12, tzinfo=tzlocal()), 'Name': 'reachability', 'Status': 'failed'}], 'Status': 'impaired'}, system_status={'Details': [{'Name': 'reachability', 'Status': 'passed'}], 'Status': 'ok'}, scheduled_events=None)
2020-12-28 15:21:01,369 - [slurm_plugin.clustermgtd:_handle_health_check] - WARNING - Setting nodes failing health check type ec2_health_check to DRAIN: ['st-c518xlarge-148']
....
2020-12-28 16:09:03,865 - [slurm_plugin.clustermgtd:_handle_health_check] - WARNING - Node st-c518xlarge-148(10.230.131.146) is associated with instance i-0c52326aac5dd766b that is failing ec2_health_check. EC2 health state: EC2InstanceHealthState(id='i-0c52326aac5dd766b', state='running', instance_status={'Details': [{'ImpairedSince': datetime.datetime(2020, 12, 28, 15, 12, tzinfo=tzlocal()), 'Name': 'reachability', 'Status': 'failed'}], 'Status': 'impaired'}, system_status={'Details': [{'Name': 'reachability', 'Status': 'passed'}], 'Status': 'ok'}, scheduled_events=None)
2020-12-28 16:09:03,865 - [slurm_plugin.clustermgtd:_handle_health_check] - WARNING - Setting nodes failing health check type ec2_health_check to DRAIN: ['st-c518xlarge-148']
2020-12-28 16:10:03,876 - [slurm_plugin.clustermgtd:_handle_health_check] - WARNING - Node st-c518xlarge-148(10.230.131.146) is associated with instance i-0c52326aac5dd766b that is failing ec2_health_check. EC2 health state: EC2InstanceHealthState(id='i-0c52326aac5dd766b', state='running', instance_status={'Details': [{'ImpairedSince': datetime.datetime(2020, 12, 28, 15, 12, tzinfo=tzlocal()), 'Name': 'reachability', 'Status': 'failed'}], 'Status': 'impaired'}, system_status={'Details': [{'Name': 'reachability', 'Status': 'passed'}], 'Status': 'ok'}, scheduled_events=None)
2020-12-28 16:10:03,876 - [slurm_plugin.clustermgtd:_handle_health_check] - WARNING - Setting nodes failing health check type ec2_health_check to DRAIN: ['st-c518xlarge-148']
2020-12-28 16:10:03,889 - [slurm_plugin.clustermgtd:_is_node_state_healthy] - WARNING - Node state check: node st-c518xlarge-148(10.230.131.146) in DRAINED, replacing node
```

node-specific events from slurm daemon:
```
[2020-12-28T15:20:00.900] update_node: node st-c518xlarge-148 reason set to: Node failing ec2_health_check                            
[2020-12-28T15:20:00.900] update_node: node st-c518xlarge-148 state set to DRAINED~                                                   
[2020-12-28T16:10:03.895] update_node: node st-c518xlarge-148 reason set to: Static node maintenance: unhealthy node is being replaced
[2020-12-28T16:10:03.895] update_node: node st-c518xlarge-148 state set to DOWN                                                       
[2020-12-28T16:15:18.059] Node st-c518xlarge-148 now responding                                                                       
[2020-12-28T16:15:18.083] update_node: node st-c518xlarge-148 reason set to: Node start up                                            
[2020-12-28T16:15:18.083] update_node: node st-c518xlarge-148 state set to IDLE                                                       
[2020-12-28T16:15:18.829] Node st-c518xlarge-148 now responding       
```

**Comment by tilne** [comment]
Hello, @ccassidy-maxar.

At first glance, this seems like an instance of a general slurm issue described here: https://github.com/aws/aws-parallelcluster/wiki/Slurm-Issues#slurm-will-not-mark-non-responsive-node-as-down-if-node-is-within-resumetimeout

However, it was previously thought that this would not apply to instances failing EC2 health checks, as noted in the 'Not Applicable' section of the linked wiki page. But in the [scontrol man page](https://slurm.schedmd.com/scontrol.html) I found the following language:
> If a node is already in the process of being powered up or down, the command will only change the state of the node but won't have any effect until the configured ResumeTimeout or SuspendTimeout is reached.

This language along with the timeline given by the log messages makes me think that the same timeout also applies even to the nodes that `clustermgtd` places into the `DRAIN` state. I'll look into this and get back to you.

In the meantime, some questions:
* Can you attach the full slurmctld and clustermgtd logs (with any confidential info redacted)?
* What was causing the status checks to fail? Are you able to reproduce the issue? If so, how?
* The wiki linked to above describes how to increase `slurmctld`'s `ResumeTimeout`. Does following those instructions result in a shorter time spent waiting for the replacement of nodes failing health checks during initialization?

**Comment by ccassidy-maxar** [comment]
Hi @tilne,
1) I have attached the 2 logs you requested
2) The status checks were failing at the system level, my assumption is bad hardware since all the other compute nodes with the same AMI worked fine.  Unable to reproduce it, and we typically see it happen a few times a month or so.
3) Unfortunately we provision a new cluster each day so not able to test with the existing cluster, I can look to implement that change and see what happens next time we get a bad node.

Thanks!

[slurmctld.log](https://github.com/aws/aws-parallelcluster/files/5751205/slurmctld.log)
[clustermgtd.log](https://github.com/aws/aws-parallelcluster/files/5751206/clustermgtd.log)

**Comment by tilne** [comment]
Thanks for the logs. I've gone through them, and also reproduced this problem by forcing status checks to fail via running `ifconfig eth0 down` on compute nodes in a post-install script. My initial workaround recommendation was incorrect. Setting a shorter ResumeTimeout didn't have any effect on how quickly the static instance failing initialization was terminated and replaced. Instead, changing clustermgtd's `node_replacement_timeout` parameter is what was needed. That said, we do recommend always setting this value and ResumeTimeout to the same. Following is the post-install script I was using to set both of them to 20 minutes (1200 seconds).

```bash
#!/bin/bash

set -ex
source /etc/parallelcluster/cfnconfig
if [ "${cfn_node_type}" = "MasterServer" ]; then
    sed -i 's/ResumeTimeout=3600/ResumeTimeout=1200/' /opt/slurm/etc/slurm.conf
    systemctl restart slurmctld
    # node_replacement_timeout isn't in the default clustermgtd config file. It defaults to 3600.
    echo >> /etc/parallelcluster/slurm_plugin/parallelcluster_clustermgtd.conf 'node_replacement_timeout = 1200'
fi
```

**Comment by ccassidy-maxar** [comment]
Hi @tilne,

Thank you for the investigation into that setting, out of curiousity what other impacts does that `node_replacement_timeout` value have on the cluster?  We spin up ephemeral clusters as needed and at least recently see the nodes go bad right away so a quicker replacement would be helpful for our use case.

Thanks!

**Comment by rexcsn** [comment]
Hi @ccassidy-maxar,

`node_replacement_timeout` only affects static nodes and is setup to mirror the behavior of `ResumeTimeout` for dynamic nodes.
Essentially both of these parameter should control the max amount of time that is allowed for instance bootstrap/setup actions.

By default, both `node_replacement_timeout` and `ResumeTimeout` are set to 3600s today to try to accommodate the time required to setup different instance types for different use cases.
However, as you pointed out, failed instance status check that occurred during instance setup are not handled correctly for static nodes during `node_replacement_timeout`.
In addition, [there is another issue](https://github.com/aws/aws-parallelcluster/wiki/Slurm-Issues#slurm-will-not-mark-non-responsive-node-as-down-if-node-is-within-resumetimeout) that caused some failures with dynamic nodes not being handled correctly during `ResumeTimeout`.
We will evaluate internally to see how we can fix both of these issues, and replace failed instances faster in general.

For your specific use case, feel free to decrease `node_replacement_timeout` and `ResumeTimeout` to a time that is safe for instances to spin up. For example, if you observed that instance setup usually take ~5 mins for your specific cluster, setting `node_replacement_timeout` and `ResumeTimeout` to 10 mins should be safe.

Hope that helps!

**Comment by ccassidy-maxar** [comment]
Thanks @rexcsn, this is excellent info!  Will implement this in our process and let you know how it reacts next time we get a bad node on spinup.

**Comment by rexcsn** [comment]
Update: this fix is not included as part of v2.10.2 patch release. We will update here when the fix is released in a future version.
Thank you.

**Comment by rexcsn** [comment]
pcluster v2.11.0 is released, which contains fix for this issue.
Resolving the issue now, please let us know if you have any question, thanks!
